README.md

```diff
@@ -4,18 +4,42 @@ Libraries making it easy to access astronomy data commons resources.
 
 ## Developer notes
 
+### Setup
+
+To prepare for development, run `pip install --editable ".[dev]"` from within
+the repo directory. This will install all dependencies, including those using
+during development workflows.
+
+This project expects you to use a `pip`-centric workflow for development on the
+project itself. It gets released to conda, but we don't attempt to record
+development-time dependencies like linters through conda tools.
```

Comment:
This text confused me, since we still have `conda-requirements.txt`.

How about something like: 
> "This project expects you to use a `pip`-centric workflow for development on the project itself. While we provide a list of `conda` packages required for development in `conda-requirements.txt`, it's not guaranteed to always be up-to-date as it's not our core workflow."
(or similar)?

Reply:
Oops. I would just like to delete `conda-requirements.txt`. If you're using conda, you should be able to use the conda environment's `pip` to `pip install -e .[dev]` on this project and get stuff done in exactly the same way as the "pip-centric" workflow described above.

Reply:
Okay.... I usually practice social distancing from `pip` due to its tendency to install dependencies and overwrite packages w/o asking, but I'm willing to give it a try!

PS: I may just be using it wrong -- does anyone know how to get pip to give me a list of everything it will do to my system and ask me to confirm before proceeding (like yum, apt, conda, ... do by default)?

Reply:
> Okay.... I usually practice social distancing from pip due to its tendency to install dependencies and overwrite packages w/o asking, but I'm willing to give it a try!

I think the most important thing is to only use `pip` within a virtualenv.

For what it's worth, while I'm doing development, I'm doing everything inside docker containers. I get some reassurance from the fact that I can blow away the container at any time.

I do that with some zsh scripts I wrote:
```zsh
function dev-docker-name {
    echo $(basename $(git rev-parse --show-toplevel))
}

function dev-docker-init {
    echo "FROM python:latest
RUN apt-get update -y && apt-get install -y zsh
COPY . /opt/$(dev-docker-name)-init
RUN pip install -e /opt/$(dev-docker-name)-init[dev]
"
}

function dev-docker-build {
    docker build \
           --tag $(dev-docker-name) \
           --file localdev.Dockerfile \
           .
}

function dev-docker-shell {
    docker run \
           --tty \
           --interactive \
           --volume $PWD:/opt/$(dev-docker-name) \
           --workdir /opt/$(dev-docker-name) \
           --rm \
           $(dev-docker-name) \
           /bin/sh
}

function dev-docker-exec {
    docker run \
           --tty \
           --interactive \
           --volume $PWD:/opt/$(dev-docker-name) \
           --workdir /opt/$(dev-docker-name) \
           --rm \
           $(dev-docker-name) \
           "$@"
}
```

To work on a project, I run make a Dockerfile with `dev-docker-init`, and then build it:
```zsh 
$ dev-docker-init > localdev.Dockerfile
$ cat localdev.Dockerfile
FROM python:latest

RUN apt-get update -y && apt-get install -y zsh
COPY . /opt/genesis-client-init
RUN pip install -e /opt/genesis-client-init[dev]

$ dev-docker-build
```

When dependencies change, I can rebuild with another call of `dev-docker-build`.

To run project stuff, I can do things like this:

Run make directives:
```
$ dev-docker-exec make lint
<... lint output ...>
$ dev-docker-exec make test
<... test output>
```

Open an interpreter:
```
-> % dev-docker-exec ipython
Python 3.8.2 (default, Feb 26 2020, 14:58:38) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import genesis.streaming                                     

In [2]: genesis.streaming.AlertBroker?                               
Init signature:
genesis.streaming.AlertBroker(
    broker_url,
    mode='r',
    start_at='latest',
    format='avro',
    auth=None,
    metadata=False,
    config=None,
)
Docstring:      <no docstring>
File:           /opt/genesis-client/genesis/streaming.py
Type:           type
Subclasses:     

In [3]:        
```

And everything is nicely isolated - nothing is installed to my laptop, so everything should work just the same in container or CI.

----

I have a ~/.gitignore line to globally ignore `localdev.Dockerfile` so that I don't force this workflow on others, but I like it. I am not prescribing it because
 1. nobody else in Python-land seems to do it, and
 2. I haven't been using it for very long

but I hope it works out.

Reply:
Along those lines -- I've been trying to kick my 25yr-long [joe](https://joe-editor.sourceforge.io/) habit and start using VS Code; I understand it has some nifty remote/container execution plugins as well. Hope to try it out soon...

Reply:
Yeah, everybody seems to use VS Code these days, but you can pry emacs from my cold dead hands :) I was thinking of writing elisp to make this workflow nice from within emacs, but only if this flow proves to be worth it for a little while.